### PR TITLE
driver: reset: npcx: add driver support for reset controller

### PR DESCRIPTION
**Pull request by ChiHuaL**
Nuvoton NPCX chips have reset registers which allow to reset the peripheral hardware modules.
This commit adds the support by implementing the reset driver.
Note that only the reset_line_toggle API is supported because of the nature of the reset controller's design.

### CODE REVIEW

File: drivers/reset/reset_npcx.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+static int reset_npcx_line_toggle(const struct device *dev, uint32_t id)
+{
+	const struct reset_npcx_dev_config *const config = dev->config;
+	struct swrst_reg *const reg = config->reg_base;
+	unsigned int key;
+	uint8_t reg_offset;
+	uint8_t reg_bit;
+
+	reg_offset = NPCX_RESET_CTL_REG_OFFSET(id);
```
Comment (keith-zephyr) [inline drivers/reset/reset_npcx.c:35]: `id` needs some bounds checks.

File: include/zephyr/dt-bindings/reset/npcx4_reset.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_RESET_NPCX4_RESET_H
+#define ZEPHYR_INCLUDE_DT_BINDINGS_RESET_NPCX4_RESET_H
+
+#define NPCX_RESET_SWRST_CTL1_OFFSET (0)
+#define NPCX_RESET_SWRST_CTL2_OFFSET (32)
+#define NPCX_RESET_SWRST_CTL3_OFFSET (64)
+#define NPCX_RESET_SWRST_CTL4_OFFSET (96)
```
[the remark this answers is not in the document]
Comment (ChiHuaL) [reply include/zephyr/dt-bindings/reset/npcx4_reset.h]: Done.

File: drivers/reset/reset_npcx.c [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+static int reset_npcx_line_toggle(const struct device *dev, uint32_t id)
+{
+	const struct reset_npcx_dev_config *const config = dev->config;
+	struct swrst_reg *const reg = config->reg_base;
+	unsigned int key;
+	uint8_t reg_offset;
+	uint8_t reg_bit;
+
+	if (id < NPCX_RESET_ID_START || id > NPCX_RESET_ID_END) {
```
Comment (fabiobaltieri) [inline drivers/reset/reset_npcx.c:43]: you can use the `IN_RANGE` macro if you want to make this a bit more compact

**Review by fabiobaltieri — DISMISSED** [review]

File: drivers/reset/reset_npcx.c (hunk 3 above)
Comment (ChiHuaL) [reply drivers/reset/reset_npcx.c]: done.

File: drivers/reset/reset_npcx.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+	key = irq_lock();
+
+	reg->SWRST_CTL[reg_offset] |= BIT(reg_bit);
+	reg->SWRST_TRG = NPCX_SWRST_TRG_WORD_CLEAR;
+	reg->SWRST_TRG = NPCX_SWRST_TRG_WORD_START;
+
+	while (reg->SWRST_TRG != NPCX_SWRST_TRG_WORD_DONE) {
+		;
+	}
```
Comment (keith-zephyr) [inline drivers/reset/reset_npcx.c:58]: Non-blocking, can you add a `WAIT_FOR` expression here as a follow up?

**Review by keith-zephyr — DISMISSED** [review]

File: drivers/reset/reset_npcx.c (hunk 4 above)
Comment (ChiHuaL) [reply drivers/reset/reset_npcx.c]: Done.

**Review by fabiobaltieri — APPROVED** [review]

**Review by MulinChao — APPROVED** [review]

**Pull request merged**